### DISCUSSION TITLE
Output in `stdout` from the `capsys` fixture is different depending on whether or not `-s` is used

### DISCUSSION THREAD

**Discussion by dgard1981** [pytest-dev/pytest#11203] [Q&A] [answered]
I'm using `pytest with the `capsys` fixture to check text output to `stdout` is as expected.

My issue is that when I run tests using the `-s` option, the assertions passes. But when I don't use `-s`, the assertion fails because some of the text in `stdout` is wrapped, meaning the actual text is no longer the same as the expected text.

Is there a way of making the `capsys` fixture capture `stdout` in the same way regardless of whether or not `-s` is used?

Note that I have also asked this question on [StackOverflow](https://stackoverflow.com/questions/76671641/make-pytest-capsys-fixture-treat-stdout-the-same-regardless-of-whether-or-not), but so far no one has been able to help.

### Example code

```python
import argparse
import pytest


class Args:

    class _HelpFormatter(argparse.HelpFormatter):

        def __init__(self, prog: str) -> None:
            super().__init__(prog, max_help_position=50)

    def __init__(self) -> None:

        global_parser = argparse.ArgumentParser(add_help=False, prog='wwaz',
            formatter_class=self._HelpFormatter)
        global_group = global_parser.add_argument_group('Global Options')
        global_group.add_argument('--help', action='help',
            help='Show this help message and exit.')
        global_group.add_argument('--verbose', action='store_true',
            default=False, help='Whether to output verbose logs. Also whether to show verbose test output. (default: False)')

        self.parser = argparse.ArgumentParser(add_help=False, prog='wwaz',
            formatter_class=self._HelpFormatter, parents=[global_parser], 
            description='Run routine Azure actions.')

    @classmethod
    def parse(self, args: list=None) -> argparse.Namespace:
        return Args().parser.parse_args(args)


@pytest.fixture(scope='module')
def expected():
    return (
        'usage: wwaz [--help] [--verbose]'
        '\n\nRun routine Azure actions.'
        '\n\nGlobal Options:'
        '\n  --help     Show this help message and exit.'
        '\n  --verbose  Whether to output verbose logs. Also whether to show verbose test output. (default: False)\n'
    )

help_args = [pytest.param(['--help'], id='--help')]

@pytest.mark.parametrize('input_args', help_args)
def test_stdout(capsys, expected, input_args):

    capsys.readouterr()
    with pytest.raises(SystemExit):
        Args.parse(input_args)

    actual, _ = capsys.readouterr()
    assert actual == expected
```

### Successful test run

Assuming the file containing the example code is ~/src/test_argparse.py -

```shell
user@computer:~/src$ pytest test_argparse.py -q -s
.
1 passed in 0.00s
```

### Unsuccessful test run

Assuming the file containing the example code is ~/src/test_argparse.py -

```shell
user@computer:~/src$ pytest test_argparse.py -q
F                                                                                                                                                                                            [100%]
============================================================================================= FAILURES =============================================================================================
_______________________________________________________________________________________ test_stdout[--help] ________________________________________________________________________________________

capsys = <_pytest.capture.CaptureFixture object at 0x7f6edf733520>
expected = 'usage: wwaz [--help] [--verbose]\n\nRun routine Azure actions.\n\nGlobal Options:\n  --help     Show this help message and exit.\n  --verbose  Whether to output verbose logs. Also whether to show verbose test output. (default: False)\n'
input_args = ['--help']

    @pytest.mark.parametrize('input_args', help_args)
    def test_stdout(capsys, expected, input_args):
    
        capsys.readouterr()
        with pytest.raises(SystemExit):
            Args.parse(input_args)
    
        actual, _ = capsys.readouterr()
>       assert actual == expected
E       AssertionError: assert 'usage: wwaz ...ult: False)\n' == 'usage: wwaz ...ult: False)\n'
E         Skipping 192 identical leading characters in diff, use -v to show
E         - rbose test output. (default: False)
E         + rbose test
E         +              output. (default: False)

test_argparse.py:51: AssertionError
===================================================================================== short test summary info ======================================================================================
FAILED test_argparse.py::test_stdout[--help] - AssertionError: assert 'usage: wwaz ...ult: False)\n' == 'usage: wwaz ...ult: False)\n'
1 failed in 0.01s
```

**Comment by The-Compiler** [comment, marked as answer]
This is unrelated to pytest - you'll see the same behaviour if you run `Args.parse()` with your code, and then look at the output of `python3 script.py --help` (output is a terminal) vs. `python3 script.py --help | cat` (output is not a terminal).

This is because `argparse.HelpFormatter` [uses](https://github.com/python/cpython/blob/v3.11.4/Lib/argparse.py#L173) `shutil.get_terminal_size()` to get a wrapping width (if none is given), which then [falls back to 80x24 when stdout is not a terminal](https://docs.python.org/3/library/shutil.html):

> When `COLUMNS` or `LINES` is not defined, which is the common case, the terminal connected to [`sys.__stdout__`](https://docs.python.org/3/library/sys.html#sys.__stdout__) is queried by invoking [`os.get_terminal_size()`](https://docs.python.org/3/library/os.html#os.get_terminal_size).
> 
> If the terminal size cannot be successfully queried, either because the system doesn’t support querying, or because we are not connected to a terminal, the value given in `fallback` parameter is used. fallback defaults to `(80, 24)` which is the default size used by many terminal emulators.

So you could either pass a width in your `_HelpFormatter`, or [use `monkeypatch`](https://docs.pytest.org/en/7.4.x/how-to/monkeypatch.html#monkeypatching-environment-variables) to set the `COLUMNS` environment variable.

**Reply by dgard1981** [reply]
Thank you for this, setting the `width` parameter for `_HelpFormatter` worked and now testing is consistent.

```python
class _HelpFormatter(argparse.HelpFormatter):

    def __init__(self, prog: str, m=25, w=120) -> None:
        super().__init__(prog, max_help_position=m, width=w) 
```
